Blog/acf (#165)
infinum/eightshift-docs: website/blog/2022-05-10-acf-in-a-project.md

---
title: ACF in a project
description: Example of using ACF in your project
slug: acf-in-a-project
authors: obradovic
date: 2022-05-10
tags: [eightshift, boilerplate, acf, advanced custom fields, theme options]
hide_table_of_contents: false
---

> If you've worked with WordPress for long enough, you've heard of the **_Advanced Custom Fields_** plugin (**_ACF_** for short). While the use of blocks has simplified content editing, making meta fields less of a necessity, there are still cases in which meta fields are very useful.

Reviewer: ```suggestion
If you've worked with WordPress for a long time, you've heard of the **_Advanced Custom Fields_** plugin (**_ACF_** for short). While the use of blocks has simplified content editing, thus making meta fields less of a necessity, there are still cases in which meta fields are very useful.
```

<!--truncate-->

## Ways of registering ACF fields

> There are multiple approaches to ACF field registration. The easiest way to register fields is using the WP Admin interface. However, if you're using Git, this may not be the best way to do it since the fields are being saved in the database, which means you would have to export and import fields across environments each time you make a change.

Reviewer: > However, if you're using Git, this may not be the best way to do it

This part makes no sense. 

Using Git has no effect on the database, or why we are using our approach.

I get what you're trying to say, but Git has nothing to do with this.

The main drawback (besides the import/export of the fields) is that when you have a large number of fields the admin can become sluggish, and fetching fields registration via database can impact performance.

Then, in the next paragraph, I'd suggest the other ways of field registration: either via PHP or using JSON, and then say that we use PHP because PHP can use opcache to speed things up even more. And then you can mention that registering fields in the PHP is better because you can commit those changes in Git for easier portability.

Reviewer 2: I believe @iobrado's point here was that ACF fields that are registered using the database can't be easily versioned using version control software. 

Your point about optimization still stands, and his point might be a bit unclear, but the idea that was communicated, from what I've gathered from the post, is that you want versioned fields for maintainability and collaboration, and field definitions as code so your theme isn't configuration-dependent.

This makes perfect sense to me, but could be a bit clearer in the post 🤷

Reviewer: I kinda see what he wanted to say, but the wording is odd/wrong so he should just reword this part and the second paragraph according to the suggestions 🙂

> We recommend registering the fields using PHP, which will allow you to version your field definitions using git. There is also an option to register the fields with **_.json_** files, but in Eightshift Development kit, we already have some goodies which will make field registration a breeze.

Reviewer: This is a part of the above comment, so just change it a bit 😄

## Getting PHP code export

> The easiest way to get the PHP code that you will use for registering ACF fields is by adding those fields through the WP Admin ACF interface first. As an example, I'll add a field group called **_Intro_**, which will contain one text field of the same name. That field group will be displayed only on posts. You can add a few more fields, but for the sake of simplicity, I'll use only one field.

Reviewer 3: I feel there is a lot of repetition of the phrase `The X way to Y is ...`
Maybe change up a bit how you start paragraphs.

> The approach we use for registering ACF fields is by adding those fields through the WP Admin ACF interface first. As an example, I'll add a field group called **_Intro_**, which will contain one text field of the same name. That field group will be displayed only on posts. You can add a few more fields, but for the sake of simplicity, I'll only use one field.

Reviewer: ```suggestion
The approach we use for registering ACF fields is by adding those fields first through the WP Admin ACF interface. As an example, We'll add a field group called **_Intro_**, which will contain one text field of the same name. That field group will be displayed only on posts. You can add a few more fields, but for the sake of simplicity, we'll only use one field.
```

When you define all the fields you need, save them and go to `Custom Fields -> Tools`. Here you will see an option to **_Export Field Groups_**. Simply select the field groups which you want to export and select **_Generate PHP_**.

![ACF PHP code export](/img/blog/acf-generate-php.png)

> This will generate a PHP code snippet that you can use in your theme. But now you may be wondering, where exactly should that code go?

Reviewer: ```suggestion
This will generate a PHP code snippet that you can use in your theme. Now you may be wondering, where exactly should that code go?
```

## CustomMeta class

> Those ACF goodies in Eightshift Development kit I talked about earlier? Let me introduce you to one of them. We have a WP CLI command which we can use to generate a CustomMeta class where we can add our field groups. The command is `wp boilerplate create_acf_meta`. This command has one required parameter, and that is `name`. To create a class that we will use for registering our custom fields, we'll use the following command:

Reviewer: ```suggestion
Those ACF goodies in Eightshift Development kit we talked about earlier? Let us introduce you to one of them. We have a WP CLI command which we can use to generate a CustomMeta class where we can add our field groups. The command is `wp boilerplate create_acf_meta`. This command has one required parameter, and that is `name`. To create a class that we will use for registering our custom fields, we'll use the following command:
```

```bash
wp boilerplate create_acf_meta --name=intro
```

This command will generate a **_CustomMeta_** folder inside **_src_** folder and add a new file called **_IntroAcfMeta.php_**. Inside that file, you should see the following method:
```php
public function fields(): void
{
	if (function_exists('acf_add_local_field_group')) {
		\acf_add_local_field_group([]);
	}
}
```

Go back to your PHP code export for ACF fields and select everything inside the `acf_add_local_field_group()` function and paste it into your function. To make it in line with our coding standards, we have to do the following:
- replace `array()` with `[]`
- make every label translatable and escaped - use `esc_html__()`

> - instead of hardcoding the field name, replace it with a class constant

Reviewer 2: Maybe explain why this is a good idea. (e.g. "to allow for easy access to field values" or "for better sustainability", etc.)

After all these changes, your code should look like this:

```php
class IntroAcfMeta extends AbstractAcfMeta
{

	/**
	 * Intro field name.
	 */
	public const INTRO_FIELD_NAME = 'intro';
	
	/**
	 * Render acf fields.
	 *
	 * @return void
	 */
	public function fields(): void
	{
		if (function_exists('acf_add_local_field_group')) {
			\acf_add_local_field_group([
				'key' => 'group_6269300acefda',
				'title' =>  \esc_html__('Intro', 'eightshift-theme'),
				'fields' => [
					[
						'key' => 'field_6269300f8029b',
						'label' => \esc_html__('Intro', 'eightshift-theme'),
						'name' => self::INTRO_FIELD_NAME,
						'type' => 'text',
						// ...
					]
				]
				// ...
			]);
		}
	}
}
```

> The final step is to go back to Custom Fields in WP Admin and delete your field group from there. Your field should now be visible when editing Posts.

Reviewer 2: ```suggestion
The final step is to go back to Custom Fields in WP Admin and delete your field group from there, to prevent registering the fields twice. After adding field definitions in PHP and removing them in WP Admin, your field should be visible when editing Posts.
```

> The final step is to go back to Custom Fields in WP Admin and delete your field group from there. Your field should now be visible when editing Posts.

Reviewer 3: ```suggestion
The final step is to go back to Custom Fields in WP Admin and either delete or deactivate your field group from there. Your field should now be visible when editing Posts.
```

> The final step is to go back to the Custom Fields in WP Admin and either delete or deactivate your field group from there, to prevent registering the fields twice. After adding field definitions in PHP and removing them in WP Admin, your field should be visible when editing Posts.

Reviewer: ```suggestion
The final step is to go back to the Custom Fields in WP admin and either delete or deactivate your field group from there, to prevent registering the fields twice. After adding field definitions in PHP and removing them in WP admin, your field should be visible when editing posts.
```

## Using get_field()

> To fetch the saved meta value, we use ACF's `get_field()` function, but here are a few tips that could improve your code quality. First, you should check if that function exists. If ACF gets for whatever reason deactivated on your site, it won't break your site. The second one is using class constant instead of hardcoding the field name. With these practices in mind, your code should look like this:

Reviewer 2: ```suggestion
To fetch the saved meta value, we use ACF's `get_field()` function, but here are a few tips that could improve your code quality. First, you should check if that function exists. That way, if ACF is deactivated on your site for whatever reason, your site won't break. The second tip is to use a class constant instead of hardcoding the field name. With these practices in mind, your code should look like this:
```

The second tip is also related to the "coding standards" checklist above, so please think of ways to connect that train of thought 😄

```php
use YourNamespace\CustomMeta\IntroAcfMeta;

// ...

if (function_exists('get_field')) {
	$introText = get_field(IntroAcfMeta::INTRO_FIELD_NAME, $postId);
}
```

> > It's better to use class constants because if you decide to change the field name, you will have to change it only in one place.

Reviewer 2: Ah, I see you've explained this here 😄

## Theme Options

> ACF's Options page has a wide array of uses and there's a very likely chance you may also need some sort of Theme Options in your project. To make the implementation of Theme Options a bit easier, we have a CLI command which generates the `ThemeOptions` class in your project. Just use the following command:

Reviewer 2: ```suggestion
ACF's Options page has a wide array of uses and it's very likely that you'll need some sort of Theme Options in your project. To make the implementation of Theme Options a bit easier, we have a CLI command which generates the `ThemeOptions` class in your project. Just use the following command:
```

I thought we should use `[]` instead of `array` though 😂

Author: It took me a while to figure out what you mean by that 😅  nice one! 😄

```bash
wp boilerplate create_theme_options
```

> This command generates a class with two methods. The first one, `createThemeOptionsPage()` creates Theme Options page and adds it to WP Admin sidebar. The second one, `registerThemeOptions()`, is what registers the fields you will have in Theme Options. Here is an example how Theme Options look after being created with WP CLI.

Reviewer 2: ```suggestion
This command generates a class with two methods. The first one, `createThemeOptionsPage()` creates a Theme Options page and adds it to the WP Admin sidebar. The second one, `registerThemeOptions()`, is what registers the fields you will have in Theme Options. Here is an example how Theme Options look after being created using `wp boilerplate`:
```

![ACF Theme Options](/img/blog/acf-theme-options.png)

> To add fields to your Theme Options, follow the steps from the **_Getting PHP code export_** section of this post and add the fields inside the `'fields' => []` array.

Reviewer 2: ```suggestion
To add fields to your Theme Options, follow the steps from the **_Exporting PHP section_** section of this post and add the fields inside the `'fields' => []` array.
```

> 
> 
> 

Reviewer: ```suggestion

```

description: Example of using ACF in your project
slug: acf-in-a-project
authors: obradovic
date: 2022-05-17
tags: [eightshift, boilerplate, acf, advanced custom fields, theme options]
hide_table_of_contents: false
---

If you've worked with WordPress for a long time, you've heard of the **_Advanced Custom Fields_** plugin (**_ACF_** for short). While the use of blocks has simplified content editing, thus making meta fields less of a necessity, there are still cases in which meta fields are very useful.
<!--truncate-->

## Ways of registering ACF fields

> There are multiple approaches to ACF field registration. The easiest way to register fields is using the WP admin interface, but this approach has a few drawbacks. If you have many fields, this can impact performance since you're fetching fields registration via the database. Another drawback is if you have multiple environments (e.g. local, staging, production), you'll have to export the fields and import them to other environments.

Reviewer: ```suggestion
There are multiple approaches to ACF field registration. The easiest way to register fields is using the WP admin interface, but this approach has a few drawbacks. If you have many fields, this can impact performance because you are registering fields dynamically (information about registered fields is stored in the database). Another drawback is if you have multiple environments (e.g. local, staging, production), you'll have to export the fields and import them to other environments.
```

> Other ways of registering ACF fields are either via PHP or JSON. We prefer to use PHP because it can use OPcache to speed things up even more. If you are using Git, this approach is even better because you can commit the PHP code used for registering fields for easier portability accross different environments. For that reason, Eightshift Development kit already has some goodies which will make field registration a breeze.

Reviewer: ```suggestion
Other ways of registering ACF fields are either via PHP (code) or by reading from a JSON file. We prefer to use PHP because you can use OPcache to speed up field registration even more. If you are using Git, this approach is even better because you can commit the PHP code used for registering fields for easier portability across different environments. For that reason, Eightshift Development kit already has some goodies which will make field registration a breeze.
```
